pm/sender_test.go

```diff
@@ -0,0 +1,258 @@
+package pm
+
+import (
+	"crypto/rand"
+	"math/big"
+	"reflect"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/accounts"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+)
+
+func TestStartSession_GivenSomeRecipientRandHash_UsesItAsSessionId(t *testing.T) {
+	sender := defaultNewSender(t)
+	recipient := ethcommon.Address{}
+	ticketParams := defaultTicketParams(t)
+	expectedSessionID := hashToHex(ticketParams.RecipientRandHash)
+
+	sessionID := sender.StartSession(recipient, TicketParams{
+		FaceValue:         big.NewInt(0),
+		WinProb:           big.NewInt(0),
+		Seed:              big.NewInt(0),
+		RecipientRandHash: ticketParams.RecipientRandHash,
+	})
+
+	if sessionID != expectedSessionID {
+		t.Errorf("expected %v to equal %v", sessionID, expectedSessionID)
+	}
+}
+
+func TestStartSession_GivenConcurrentUsage_RecordsAllSessions(t *testing.T) {
+	sender := defaultNewSender(t)
+	recipient := ethcommon.Address{}
+
+	var sessions []string
+	var sessionCount int32
+	ch := make(chan struct{})
+	for i := 0; i < 100; i++ {
+		ticketParams := defaultTicketParams(t)
+		expectedSessionID := hashToHex(ticketParams.RecipientRandHash)
+		sessions = append(sessions, expectedSessionID)
+
+		go func() {
+			sender.StartSession(recipient, TicketParams{
+				FaceValue:         big.NewInt(0),
+				WinProb:           big.NewInt(0),
+				Seed:              big.NewInt(0),
+				RecipientRandHash: ticketParams.RecipientRandHash,
+			})
+			currentCount := atomic.AddInt32(&sessionCount, 1)
+			if currentCount >= 100 {
+				ch <- struct{}{}
+			}
+		}()
+	}
+	// waiting for all session to be created
+	<-ch
+
+	for _, sessionID := range sessions {
+		_, ok := sender.sessions.Load(sessionID)
+		if !ok {
+			t.Errorf("expected to find sessionID in sender. sessionID: %v", sessionID)
+		}
+	}
+}
+
+func TestCreateTicket_GivenNonexistentSession_ReturnsError(t *testing.T) {
+	sender := defaultNewSender(t)
+
+	_, _, _, err := sender.CreateTicket("foo")
+
+	if err == nil {
+		t.Errorf("expected an error for nonexistent session")
+	}
+	if !strings.Contains(err.Error(), "unknown session") {
+		t.Errorf("expected error to contain 'unknown session' but instead got: %v", err.Error())
+	}
+}
+
+func TestCreateTicket_GivenValidSessionId_UsesSessionParamsInTicket(t *testing.T) {
+	sender := defaultNewSender(t)
+	am := sender.accountManager.(*stubAccountManager)
```

Comment:
Do you mean for example, instead of setting `saveSignRequest`, have a function like `ShouldSaveSignRequest(bool)`?

For the case of a stub I don't think we gain much by adding a function that simply wraps a boolean, that's why I chose to stick with this approach.

What are the benefits you see us having from adding these wrapper methods?

As for the type assertion, are you suggesting that `defaultNewSender` should return both sender and accountManager instead of just the sender? - I think that can definitely work. I avoided this at first because it kept all the other tests that don't need the accountManager a bit cleaner.
Do you think it's worth adding the extra return value?

Reply:
> For the case of a stub I don't think we gain much by adding a function that simply wraps a boolean, that's why I chose to stick with this approach.

I actually just realized that since we're writing tests internal to the package, we can just directly set the unexported fields of `stubAccountManager` so ignore my comment on using an exported wrapper function!

> As for the type assertion, are you suggesting that defaultNewSender should return both sender and accountManager instead of just the sender?

I was actually thinking that for this test you could manually instantiate `stubAccountManager` and `sender` so that you can directly use `stubAccountManager` within the test

Reply:
Hm yeah that's a good consideration. With that in mind I think keeping the type assertion + the current approach for instantiation is reasonable for now